### ISSUE TITLE
 WW3 multi grid mode - MPI tag overflow

### ISSUE THREAD

**Issue by haititu** [NOAA-EMC/WW3#349]
Dear WW3 folks

Has anyone else encountered MPI issues with running WW3 in multi grid mode (ww3_multi)?

Our test run failed with apparently MPI tag overflow errors as shown below.
Some articles and forums on the web suggest that these are due to MPI library implementations which may have lower upperbounds for the MPI tag values.

We used IntelMPI v2020.4.304.
The PBS script we used is appended below.

Previously we ran the exact same experiment on an older version of Intel MPI, v2015, and did not encounter these MPI errors.

Could this be a WW3 code issue where the MPI tag values generated are
too large, i.e. beyond the MPI Standard guarantee of 32K?

------------------
```
Abort(134849796) on node 48 (rank 48 in comm 0): Fatal error in
PMPI_Irecv: Invalid tag, error stack:
PMPI_Irecv(162): MPI_Irecv(buf=0x2af4a3d4d420, count=768, MPI_REAL,
src=63, tag=1296268, MPI_COMM_WORLD, request=0x18ffc8f8) failed
PMPI_Irecv(96).: Invalid tag, value is 1296268
Abort(269067524) on node 51 (rank 51 in comm 0): Fatal error in
PMPI_Irecv: Invalid tag, error stack:
PMPI_Irecv(162): MPI_Irecv(buf=0x2af916f73420, count=768, MPI_REAL,
src=189, tag=1092834, MPI_COMM_WORLD, request=0x198c8768) failed
PMPI_Irecv(96).: Invalid tag, value is 1092834
Abort(805938436) on node 52 (rank 52 in comm 0): Fatal error in
PMPI_Irecv: Invalid tag, error stack:
PMPI_Irecv(162): MPI_Irecv(buf=0x2baab7e6e020, count=768, MPI_REAL,
src=217, tag=1093675, MPI_COMM_WORLD, request=0x180b48a4) failed
PMPI_Irecv(96).: Invalid tag, value is 1093675
```
------------------
```

#!/bin/bash
#PBS -N long_test
#PBS -q workq
#PBS -l select=6:ncpus=48:mpiprocs=48:mem=128gb
#PBS -l walltime=36:00:00

cd /home/WW3/RESEARCH/WWATCH3/model

module load parallel_studio_xe_2020
module load netcdf-c-4.7.4

mpirun -np `cat $PBS_NODEFILE | wc -l` exe/ww3_multi

```
------------------

regards
chk

**Comment by JessicaMeixner-NOAA** [comment]
We have seen the MPI tag issues in WW3 in the past.  Is there any chance you can add the trace back to see which MPI call you are running into this issue on?

**Comment by haititu** [comment]
Hi Jessica
Sorry for my late reply.
Unfortunately, we weren't able to get the trace back.
However, we managed to implement a workaround for this issue by setting environment variables:

MPIR_CVAR_CH4_OFI_TAG_BITS=31
MPIR_CVAR_CH4_OFI_RANK_BITS=8

as described here:

https://doku.lrz.de/display/PUBLIC/Intel+MPI
under "MPI Tag's Upper bound exceeded"

But note the caveats.

We also realised we failed to RTFM. :)  In the WW3 manual, in section "C.3 MPI point-to-point communication errors", it describes this MPI tag overflow as a known issue and its workaround, but we have not tried them.

We'd be interested to hear any further comments to this issue. 😀

regards
chk
